### DISCUSSION TITLE
Business rules with lower priority (-10) being called/added

### DISCUSSION THREAD

**Discussion by Chicagoan2016** [MarimerLLC/csla#3366] [Questions]
This is related to #3365 
We have assigned priorities to business rules, we don't set ProcessThroughPriority in our code.

I have two rules, one rule (EmployeeIDgreaterzero) with priority of -10 checks that EmployeeID is greater than zero (EmploeeID being an int property).
If not, it calls context.AddErrorResult with a error message something like this
 ```cs
              context.AddErrorResult(message);
  ```
The other rule has priority of 0 and takes two properties, EmployeeID and departmentname.
We named the rule as DepartmentexistsforEmployee.
The DepartmentName is string property with default of String.Empty (we provide the default value in declaration, csla property syntax).

When this line is reached in AddBusinessRules
 ```cs
               BusinessRules.AddRule(new DepartmentexistsforEmployee.....
  ```
It causes the 
       base.DataPortal_Create() to throw exception of null object as discussed in #3365

**Comment by Chicagoan2016** [comment]
Now all this is related to #3361  :)
If we look at the following

```cs
private IPropertyInfo DepartmentName { get; set; }

public DepartmentExistsforEmployee(IPropertyInfo primaryProperty, IPropertyInfo departmentname)
          : base(primaryProperty)
        {
    InputProperties.Add(PrimaryProperty);
    InputProperties.Add(departmentname);
    //If I change the above line to 
    // InputProperties.Add(DepartmentName);
    // I am getting the errors mentioned in #3365 
}
```

The above rule is inherited from CommonBusinessRule

**Comment by Chicagoan2016** [comment]
I suppose I read it wrong, This is @jonnybee 's code
```cs
   public LessThan(IPropertyInfo primaryProperty, IPropertyInfo compareToProperty)
      : base(primaryProperty)
    {
      CompareTo = compareToProperty;
      InputProperties = new List<IPropertyInfo>() { primaryProperty, compareToProperty };
      this.RuleUri.AddQueryParameter("compareto", compareToProperty.Name);
    }
```

**Comment by rockfordlhotka** [comment]
The data portal should not be involved in the creation of a rule object. If `new MyRule()` causes a data portal exception, that seems to indicate that the rule's constructor is calling the data portal? That'd be bad!

**Reply by Chicagoan2016** [reply]
@rockfordlhotka , This particular business rule is not calling a command object inside the Execute method but our other business rules do.
The following scenario is causing the error, which seems somewhat strange.
```cs
      private IPropertyInfo CompareTo { get; set; }
    public LessThan(IPropertyInfo primaryProperty, IPropertyInfo compareToProperty)
      : base(primaryProperty)
    {
  
      CompareTo = compareToProperty;
        //The following works fine
      InputProperties = new List<IPropertyInfo>() { primaryProperty, compareToProperty };
     
     //If I do this, base.DataPortal_Create() throws an error, object not set to an instance......
    // Here I am using the Property names as declared in the rule object.
    //InputProperties = new List<IPropertyInfo>() { PrimaryProperty, CompareToProperty };
   

      this.RuleUri.AddQueryParameter("compareto", compareToProperty.Name);
    }

```
